sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/MetadataColumn.java

```diff
@@ -36,6 +36,45 @@
  */
 @Evolving
 public interface MetadataColumn {
+  /**
+   * Indicates whether a row-level operation should preserve the value of the metadata column
+   * for deleted rows. If set to true, the metadata value will be retained and passed back to
+   * the writer. If false, the metadata value will be replaced with {@code null}.
+   * <p>
+   * This flag applies only to row-level operations working with deltas of rows. Group-based
+   * operations handle deletes by discarding matching records.
+   *
+   * @since 4.0.0
+   */
+  String PRESERVE_ON_DELETE = "__preserve_on_delete";
+  boolean PRESERVE_ON_DELETE_DEFAULT = true;
+
+  /**
+   * Indicates whether a row-level operation should preserve the value of the metadata column
+   * for updated rows. If set to true, the metadata value will be retained and passed back to
+   * the writer. If false, the metadata value will be replaced with {@code null}.
+   * <p>
+   * This flag applies to both group-based and delta-based row-level operations.
+   *
+   * @since 4.0.0
+   */
+  String PRESERVE_ON_UPDATE = "__preserve_on_update";
+  boolean PRESERVE_ON_UPDATE_DEFAULT = true;
+
+  /**
+   * Indicates whether a row-level operation should preserve the value of the metadata column
+   * for inserts generated by splitting updated rows into deletes and inserts. If true,
+   * the metadata value will be retained and passed back to the writer. If false, the
+   * metadata value will be replaced with {@code null}.
+   * <p>
+   * This flag applies only to row-level operations working with deltas of rows. Group-based
+   * operations do not represent updates as deletes and inserts.
+   *
+   * @since 4.0.0
+   */
+  String PRESERVE_ON_INSERT_AS_UPDATE = "__preserve_on_insert_as_update";
```

Comment:
In theory, we can get rid of this separate flag and respect `PRESERVE_ON_UPDATE` when updates are split. That said, it will be a behavior change compared to the existing implementation that always discards metadata columns when updates are split into deletes and inserts. The API is marked as `@Experimental` but it probably makes sense to be safe. Let me know what everybody thinks.

Reply:
And for direct "insert" we always preserve the metadata columns?

Reply:
Metadata column values for new records are always `null` as such records come from input. I added this to the description above.

Reply:
Actually, let me reconsider the naming a bit more here. Please ignore for now.

Reply:
Resolving, no longer applies as the code was changed.